pageserver/src/keyspace.rs

```diff
@@ -61,6 +62,60 @@ impl KeySpace {
 
         KeyPartitioning { parts }
     }
+
+    /// Add range to keyspace.
+    ///
+    /// Unlike KeySpaceAccum, it accepts key ranges in any order and overlapping ranges.
+    pub fn add_range(&mut self, range: Range<Key>) {
```

Comment:
O, sorry - it is problem of replacing BTreeMap with binary search.
I should definitely add such test myself.
Now it is passed.

> Also, if you add a range that overlaps with multiple existing ranges, it doesn't merge them correctly, leaving overlapping ranges in the KeySpace.

There is a loop which should collapse all overlapping ranges

> We don't actually need to maintain a valid KeySpace during GC.

Yes, it is true.
But sorting overlapped regions is non-trivial: we in any case need to sort them by start or end and then implement some logic of collapsing overlapped ranges. It will be not simpler than `add_range` function above. But certainly in this case we do not need to use binary search for each insertion. But we do need to perform sort. So final complexity will be the same : N*log(N). I prefer to keep thinks simple and preserve KeySpace invariant.

Reply:
Here's another failing case:
```
diff --git a/pageserver/src/keyspace.rs b/pageserver/src/keyspace.rs
index 131300e79..aa187789c 100644
--- a/pageserver/src/keyspace.rs
+++ b/pageserver/src/keyspace.rs
@@ -298,11 +298,22 @@ mod tests {
         //
         // #####   #####
         //    #######
-        println!("joins");
         let mut ks = KeySpace::default();
         ks.add_range(kr(0..10));
         ks.add_range(kr(20..30));
         ks.add_range(kr(5..25));
         assert_ks_eq(&ks, vec![kr(0..30)]);
+        
+        // add a range that overlaps with two existing ranges, joining them
+        //
+        // #####   #####   #####   #####
+        //    #######################
+        let mut ks = KeySpace::default();
+        ks.add_range(kr(0..5));
+        ks.add_range(kr(10..15));
+        ks.add_range(kr(20..25));
+        ks.add_range(kr(30..35));
+        ks.add_range(kr(8..32));
+        assert_ks_eq(&ks, vec![kr(0..35)]);
     }
 }
```

Reply:
>So final complexity will be the same : N*log(N).

Note that when you insert an element into an array, you need to move all the elements that follow, to make room for the new element. Similarly when you remove an element. So in the worst case that you insert n element in reverse order, I think this is O(n^2).

I'm not too worried about that worst case in practice, but I would prefer a collect and sort approach. It would avoid the O(n^2) worst case altogether, and it would be simpler and easier to convince myself that it's correct.

Reply:
Please notice that LayerMap is using BTreeMap inside to store set of historic layers ordered by key range start.
It is exactly the same order which we need for KeySpace.
I am not sure that we always traverse layers in start key range increasing order, but in most cases it is definitely true.
So we do not to perform massive copies of array elements - most frequently we will just append  items to this vector.
From my point of view having ad_range method in KeySpace is better because:
1. It simplifies code - no need to create yet another builder class.
2. It reduce size of this array (because overlapped range will be immediately merged)
But if you insist, I can rewrite  this code to use  KeySpaceRandomAccum

Reply:
Yeah, I would much prefer a KeySpaceRandomAccum. I wouldn't trust myself writing `add_range` correctly :-).

Reply:
Ok, done